### PR TITLE
Prefix temporary variable names with function id

### PR DESCRIPTION
**Pull request by peterschrammel**
Temporary variable names must be globally unique and
changing the number of temporaries in one function
must not change a variable name in another function.
Otherwise, this has adverse effects on goto-diff.

This PR gets rid of various global temporary
variables counters that do not meet above criteria.
Instead, we now always use the function id to prefix
temporaries to eliminate cross-function effects
and increment the numeric suffix only when the
entire variable name already exists, using the rename 
facility that has already been used for the same purpose 
in other places.

### CODE REVIEW

File: regression/cbmc/pointer-function-parameters-2/test.desc [hunk 1]
```diff
@@ -3,10 +3,10 @@ main.c
 --function fun --cover branch
 ^\*\* 7 of 7 covered \(100.0%\)$
 ^Test suite:$
-^a=\(\(signed int \*\*\)NULL\), tmp\$\d+=[^,]*, tmp\$\d+=[^,]*$
-^a=&tmp\$\d+!0, tmp\$\d+=\(\(signed int \*\)NULL\), tmp\$\d+=[^,]*$
-^a=&tmp\$\d+!0, tmp\$\d+=&tmp\$\d+!0, tmp\$\d+=(-[0-9]+|[012356789][0-9]*|4[0-9]+)$
-^a=&tmp\$\d+!0, tmp\$\d+=&tmp\$\d+!0, tmp\$\d+=4$
+^a=\(\(signed int \*\*\)NULL\), tmp(|\$\d+)=[^,]*, tmp(|\$\d+)=[^,]*$
+^a=&tmp(|\$\d+)!0, tmp(|\$\d+)=\(\(signed int \*\)NULL\), tmp(|\$\d+)=[^,]*$
+^a=&tmp(|\$\d+)!0, tmp(|\$\d+)=&tmp(|\$\d+)!0, tmp(|\$\d+)=([012356789][0-9]*|4[0-9]+)$
```
Comment (tautschnig) [inline regression/cbmc/pointer-function-parameters-2/test.desc]: I believe this lost the change from 5832ffd19edf0ca0cab8ea25e8ce79474083530d

File: regression/cbmc/Quantifiers-type/test.desc [hunk 2]
```diff
@@ -2,8 +2,8 @@ CORE
 main.c
 
 ^\*\* Results:$
-^\[main.assertion.1\] assertion tmp_if_expr\$1: FAILURE$
-^\[main.assertion.2\] assertion tmp_if_expr\$2: SUCCESS$
+^\[main.assertion.1\] assertion tmp_if_expr(|\$\d+): FAILURE$
```
Comment (tautschnig) [inline regression/cbmc/Quantifiers-type/test.desc]: I think `(\$\d+)?` is more idiomatic than `(|...)`

**Review by tautschnig — CHANGES_REQUESTED** [review]
I believe this is very helpful, thinking about the code that dump-c happily generates. Just one problem (as noted below): 5832ffd19edf0ca0cab8ea25e8ce79474083530d got lost.

**Comment by peterschrammel** [comment]
@tautschnig, I've addressed your comments.

**Review by tautschnig — APPROVED** [review]

**Review by NlightNFotis — APPROVED** [review]
LGTM

File: src/java_bytecode/java_static_initializers.cpp [hunk 3]
```diff
@@ -114,6 +114,7 @@ static void create_clinit_wrapper_symbols(
   symbolt wrapper_method_symbol;
   code_typet wrapper_method_type;
   wrapper_method_type.return_type() = void_typet();
+  wrapper_method_type.parameters(); // to ensure the property is there
```
Comment (tautschnig) [inline src/java_bytecode/java_static_initializers.cpp]: Why is this necessary?! The only impact that I can think of is the hash value and comparison between irepts.

Comment (peterschrammel) [reply src/java_bytecode/java_static_initializers.cpp]: Exactly for that reason.

Comment (tautschnig) [reply src/java_bytecode/java_static_initializers.cpp]: So what code is it that is impacted by changes in the ordering resulting from those differences here?

Comment (peterschrammel) [reply src/java_bytecode/java_static_initializers.cpp]: goto-diff

Comment (tautschnig) [reply src/java_bytecode/java_static_initializers.cpp]: But then the same program should always yield the same `irept` (previously without an `ID_parameters` in `named_sub` and now with such an entry). Hence I'm still surprised that this impacts goto-diff in some way.

Comment (peterschrammel) [reply src/java_bytecode/java_static_initializers.cpp]: In principle, yes. The root cause is likely an instability elsewhere in the code. I found this issue only after diffing two 200MB irep dumps. We'll have to investigate this further by shrinking the large example that I have. Any similar sources of nondeterminism will cause further trouble in future. For now, this fix will solve the problem for the customer.

Comment (tautschnig) [reply src/java_bytecode/java_static_initializers.cpp]: You might just want to log this in an issue to ensure immediate progress while not forgetting about the problem. As the issue surfaces in goto-diff, however, which data structure in goto-diff is it that is being traversed and found to be spuriously different?

**Pull request merged**